### DISCUSSION TITLE
Switch to a newer build system?

### DISCUSSION THREAD

**Discussion by jimmywarting** [agershun/alasql#1558] [Ideas]
It was very long time ago since i saw gulp being used for it's build system.
specially so when using the very old concatinate build strategy.

what would you say about trying to switch it out for something a bit more modern like using esbuild and ESM syntax?
that way you would have better type interference.

doe it would require a very hectic transformation to go from one thing to the other, maybe come up with some plan to do so?

**Comment by mathiasrw** [comment]
Concatenating strings is an old art that never should have existed. 

We _can_ put everything into one big file and swap to esbuild (yep - the right tool) - but that would make it painful to edit - so we want to keep modules with functionality. Happy to see some progress here. I think I started 3 times over the last few years, but life and pandemic stole my focus each time. 

Because we want to support slightly different functionality in web and node we want to think about how we can pass the functions that are different as a package. I imagine having the default for the browser/deno and then having a dedicated config package for node. 

Something along the lines with 

```
import alasql from 'alasql'
import * as alasqlNodeConf from 'alasql/node-config'
alasql.config(alasqlNodeConf)
```

Or maybe just publish a wrapper package that does it for you so you end up just doing a `import alasql from 'alasql/node'` in node.

There are many ways to do it - but we just need to be smart with this so we can keep expanding without changing the ground structure. 

Another thing to consider is how best to make the jison output (the code reading the string inputs) to fit into this structure.

**Reply by jimmywarting** [reply]
it would be neat if it were more modular so you could import just the things that you needed, not just solo depend on treeshaking.
CDN still do not have the same compatibility, so if you could do something more in lines of:
```js
import base from 'alasql'
import idb from 'alasql/indexeddb.js'
```
than i think that would be 👍 

put everything into one big file would probably be the easiest step forward. taking baby/incremental steps.

**Reply by jimmywarting** [reply]
I see potentials that do not involves concat everything into one file. 
If you just let me work on each file a little bit then i can gradually improve stuff. 
just need to get familiar to the code stack a bit first.

**Reply by jimmywarting** [reply]
here is my idea: 
- Move all kind of exports to the bottom of the file.<br> ( That makes it more clear of what's being exported and what is available )
- carefully add brackets around certain files when needed and use `let`, `const` or `class`  instead of `var`s <br>( As `var` host variable upwards )<br> ( Using brackets makes it more like a module - it's similar to a anonymous iffy but less verbose )
- Add only named ESM exports to the bottom of the file right underneath the alasql exports 
- If a file depends on something, use `import` at the top of the file.
- Before doing all of this we have to add a gulp step that removes import / export statements so that it dose not end up in the dist
- (I also do not think that there should be any default exports...)


So `21transaction.js` would kind of look like going from this: 
```js
var Transaction = (alasql.Transaction = function (databaseid) {
  // ...
}
```
into this:
```js

class Transaction {
  // ...
}

/***********************************************/
/*                   Exports                   */
/***********************************************/

alasql.Transaction = Transaction;

export { Transaction }
```
`export { Transaction }` would get stripped out for now... 
then eventually we would be able make the switch to remove `alasql.Transaction = Transaction;` instead.

**Reply by jimmywarting** [reply]
Maybe a bit weird and complicated. perhaps concating files is easier... doe it would help alot if the concatinated file included start end of file {x} comments of each insertion

**Reply by jimmywarting** [reply]
I notice that one thing that helps alot is to add `/** @namespace alasql */` to each file atm...

**Reply by mathiasrw** [reply]
Sounds really neat. What do you need from me to get started with wrapping into classes?

We might need a specific way to wrap the parser as its autogenerated from jison.

**Reply by jimmywarting** [reply]
I'm not 100% sure yet. as i keep modernizing the code bit, by bit then i learn more and more of how to best come up with a good solution.

**Reply by leqwasd** [reply]
I would even argue that some of alasql's "features" shouldn't even exist here, like - xlsx integration, CLI, specific node calls - those could be moved to a separate library, like, 'alasql/xlsx', 'alasql-cli', 'alasql-node'

**Reply by mathiasrw** [reply]
Yep. Having it modular would make it easier to look into. 

I did dive into how to take the current code and find a way to put it into modules - without rewriting every single thing.

Is this a sane way to move forward? What could be improved in terms of the ground structure and build system? https://github.com/AlaSQL/alasql/tree/feature/rollup

**Reply by leqwasd** [reply]
I would like to contribute there, but.. It is currently ~451 commits behind develop :(

**Reply by mathiasrw** [reply]
Hah. Yes. We need to recreate the efforts (wellll - actually there is a lot of manual work in converting the files, so it might not be so bad to copy paste the code from the new files instead of the old ones). 

First of all, lets find a good philosophy for how we want the repo to work. I am here thinking of the battle between commonjs and the es6 modules, what we have access to of new build tools, and how we would like things to work. 

I propose a proof of concept - it does not need to provide the right data - but just compile something we can import and talk about.

**Reply by leqwasd** [reply]
And how do we approach code splitting?- monorepo?

**Reply by mathiasrw** [reply]
Code splitting?

**Reply by mathiasrw** [reply]
I am open to all sorts of inputs. If anyone got good experience with now things can flow smootly im happy to hear. Lets get this thing in better shape.

**Comment by mathiasrw** [comment]
I moved the discussion from https://github.com/AlaSQL/alasql/issues/1240 to this page.

**Comment by mathiasrw** [comment]
It has been decided that we will transition to use Bun for building and testing. 

The work will be picked up in the mwulff/esm branch, and will after passing all tests live in "main" branch that will be made the default branch instead of develop.

**Reply by mathiasrw** [reply]
Getting back onto this in December.
